lms/extractors/ziparchive.py

```diff
@@ -30,35 +30,65 @@ def can_extract(self) -> bool:
         return self.is_zipfile
 
     @staticmethod
-    def _extract(archive: ZipFile, filename: str) -> File:
+    def _extract(archive: ZipFile, filename: str, dirname: str = '') -> File:
         with archive.open(filename) as current_file:
             log.debug(f'Extracting from archive: {filename}')
             code = current_file.read()
         decoded = code.decode('utf-8', errors='replace').replace('\x00', '')
+        filename = filename[len(dirname):]
         return File(path=f'/{filename}', code=decoded)
 
-    def get_exercise(self, file: ZipFile) -> Tuple[int, List[File]]:
+    def get_files(
+        self, archive: ZipFile, filenames: List[Text], dirname: str = '',
+    ) -> Iterator[File]:
+        unwanted_files = self.get_unwanted_files(filenames)
+        yield from(
+            self._extract(archive, filename, dirname)
+            for filename in filenames
+            if (
+                filename.startswith(dirname)
+                and filename not in unwanted_files
+                and filename != dirname
+            )
+        )
+
+    def get_exercises_by_dirs(
+        self, archive: ZipFile, filenames: List[Text],
+    ) -> Iterator[Tuple[int, List[File]]]:
+        for dirname in filenames:
+            parent_name, _ = os.path.split(dirname)
+            exercise_id, _ = self._clean(parent_name)
+            if (
+                exercise_id
+                and len(dirname.strip(os.path.sep).split(os.path.sep)) == 1
```

Comment:
Put it in a variable and get the condition to be in one line

Reply:
I put another condition in the beginning of the for loop in order to check in advance the directories and to be more efficient